src/CodeGen_D3D12Compute_Dev.cpp

```diff
@@ -957,6 +958,46 @@ void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::visit(const Atomic *op) {
     user_assert(false) << "Atomics operations are not supported inside D3D12Compute kernel.\n";
 }
 
+void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::visit(const Shuffle *op) {
+    if (op->type.is_scalar()) {
+        CodeGen_C::visit(op);
+    } else {
+        internal_assert(!op->vectors.empty());
+        for (size_t i = 1; i < op->vectors.size(); i++) {
+            internal_assert(op->vectors[0].type() == op->vectors[i].type());
+        }
+        internal_assert(op->type.lanes() == (int)op->indices.size());
+        const int max_index = (int)(op->vectors[0].type().lanes() * op->vectors.size());
+        for (int i : op->indices) {
+            internal_assert(i >= 0 && i < max_index);
+        }
+
+        std::vector<string> vecs;
+        for (const Expr &v : op->vectors) {
+            vecs.push_back(print_expr(v));
+        }
+
+        string src = vecs[0];
+        ostringstream rhs;
+        // This code has always assumed/required that all the vectors
+        // have identical types, so let's verify
+        const Type t0 = op->vectors[0].type();
+        for (const auto &v : op->vectors) {
+            internal_assert(t0 == v.type());
+        }
```

Comment:
This logic for verifying that the vector types match appears to be duplicated above (and likewise in the Metal and OpenCL backends)